File: pset6/credit/credit.py
from cs50 import get_int

def main():
    # Prompt user for credit card number
    number = get_int("Number: ")
    # Convert to a string to do some analysis
    string = str(number)
    length = len(string)
    firstTwoDigits = int(string[0:2])
    # Loop through every number from the back
    runningSum = 0
    isSpecial = False
    while number > 0:
        lastDigit = number % 10
        # Check for special algorithm, then flip the logic for the next number, and reduce that number.
        if (isSpecial):
            runningSum += calcSpecial(lastDigit)
        else:
            runningSum += lastDigit
        isSpecial = not isSpecial
        number = (number - lastDigit) / 10

    print(typeOfCard(runningSum, firstTwoDigits, length))

def calcSpecial(digit):
    digit *= 2
    if (digit >= 10):
        firstDigit = (digit - digit % 10) / 10
        secondDigit = digit % 10
        return firstDigit + secondDigit
    else:
        return digit

def typeOfCard(checksum, firstTwo, length):
    # If the checksum is wrong, it is not a card
    if (checksum % 10 != 0):
        return "INVALID"
    # If it is 15-digit 34xxx or 37xxx, it is AMEX
    if (length == 15 and (firstTwo == 34 or firstTwo == 37)):
        return "AMEX"
    # If it is 16-digit 51xx 52xx 53xx 54xx 55xx, it is MasterCard
    if (length == 16 and firstTwo >= 51 and firstTwo <= 55):
        return "MASTERCARD"
    # If it is 13 or 16 digit 4xxxxxxx it is VISA
    if ((length == 13 or length == 16) and firstTwo >= 40 and firstTwo <= 49):
        return "VISA"
    # If all those checks failed, no cardtype exists for this.
    return "INVALID"

main()

File: pset1/credit/credit.c
#include <stdio.h>
#include <cs50.h>

int main(void) {
    // Request a Number
    long creditCardNumber;
    creditCardNumber = get_long("Credit Card Number: ");
    // assumes first digit is not 0 from now

    // Create a Duplicate Number for length investigation.
    long number = creditCardNumber;
    int lengthOfNumber = 0;
    while (number != 0) {
        number = number / 10;
        lengthOfNumber++;
    }

    // Investigate first 2 digits.
    number = creditCardNumber;
    int firstTwoDigits;
    for (int i = 0; i < lengthOfNumber - 2; i++) {
        number = number / 10;
    }
    firstTwoDigits = number;

    // Get first digit for convenience.
    int firstDigit = firstTwoDigits / 10;

    // Check if the card is valid.
    number = creditCardNumber;
    int runningSum = 0;
    // Loop through every number.
    for (int i = 0; i < lengthOfNumber; i++) {
        int lastDigit = number % 10;
        // Check the digit's parity and proceed to perform their respective algorithm.
        if (i % 2 == 0) {
            // If it is even, number is taken as it is.
            runningSum += lastDigit;
        } else {
            // If it is odd, the algorithm is performed.
            lastDigit = lastDigit * 2;
            runningSum += lastDigit % 10;
            runningSum += lastDigit / 10;
        }
        // change lastDigit
        number = number / 10;
    }

    // Check if checksum is valid.
    if (runningSum % 10 == 0) {
        // It is valid.

        // American Express? if not, MasterCard? if not, VISA?
        if (lengthOfNumber == 15 && ( firstTwoDigits == 34 || firstTwoDigits == 37 )) {
            printf("%s\n", "AMEX");
        } else if (lengthOfNumber == 16 && firstTwoDigits > 50 && firstTwoDigits < 56) {
            printf("%s\n", "MASTERCARD");
        } else if ( (lengthOfNumber == 13 || lengthOfNumber == 16) && firstDigit == 4) {
            printf("%s\n", "VISA");
        } else {
            printf("%s\n", "INVALID");
        }
    } else {
        // It is invalid.
        printf("%s\n", "INVALID");
    }
}


